cranelift/codegen/src/dominator_tree.rs

```diff
@@ -242,88 +290,118 @@ impl DominatorTree {
         self.valid
     }
 
-    /// Reset all internal data structures and compute a post-order of the control flow graph.
-    ///
-    /// This leaves `rpo_number == 1` for all reachable blocks, 0 for unreachable ones.
-    fn compute_postorder(&mut self, func: &Function) {
-        self.clear();
+    /// Reset all internal data structures, build spanning tree
+    /// and compute a post-order of the control flow graph.
+    fn compute_spanning_tree(&mut self, func: &Function) {
         self.nodes.resize(func.dfg.num_blocks());
-        self.postorder.extend(self.dfs.post_order_iter(func));
+        self.stree.reserve(func.dfg.num_blocks());
+
+        if let Some(block) = func.layout.entry_block() {
+            self.dfs_worklist.push(TraversalEvent::Enter(0, block));
+        }
+
+        loop {
+            match self.dfs_worklist.pop() {
+                Some(TraversalEvent::Enter(parent, block)) => {
+                    let node = &mut self.nodes[block];
+                    // The node has been visited
+                    if node.pre_number != 0 {
+                        continue;
+                    }
+
+                    self.dfs_worklist.push(TraversalEvent::Exit(block));
+
+                    let pre_number = self.stree.push(parent, block);
+                    node.pre_number = pre_number;
+
+                    // Use the same traversal heuristics as in traversals.rs
+                    self.dfs_worklist.extend(
+                        func.block_successors(block)
+                            .rev()
+                            .filter(|successor| self.nodes[*successor].pre_number == 0)
```

Comment:
The `0` sentinel keeps occurring and I think it'd be clearer if we had a name for it -- could you define `const NOT_VISITED: u32 = 0;` or similar in this module?